src/compiler/scanner.ts

```diff
@@ -858,28 +858,76 @@ namespace ts {
             }
         }
 
+        function scanNumberFragment(): string {
+            let start = pos;
+            let allowSeperator = false;
+            let result: string;
+            while (true) {
+                const ch = text.charCodeAt(pos);
+                if (allowSeperator && ch === CharacterCodes._) {
+                    allowSeperator = false;
+                    tokenFlags |= TokenFlags.ContainsSeparator;
+                    result = (result || "") + text.substring(start, pos);
+                    pos++;
+                    start = pos;
+                    continue;
+                }
+                if (isDigit(ch)) {

```

Comment:
Considering the similarities between this, `scanHexDigit`, and `scanBinaryOrOctalDigits` perhaps we should unify (at least in part) these three functions by having `scanNumberFragment` take a callback for `isDigit` and then calling it as `scanNumberFragment(isDigit)`, `scanNumberFragment(isHexDigit)`, `scanNumberFragment(isBinaryDigit)`, and `scanNumberFragment(isOctalDigit)`.

Reply:
`scanBinaryOrOctalDigits` and `scanHexDigit` also do arithmetic to calculate the actual numeric value, however (rather than relying on the host being able to parse the literal correctly).  I would also need to pass thru another callback to retain that behavior, too.

Reply:
`scanHexDigit` also has to deal with validating that there are a certain number of digits and have an option to not recognize separators, for when its used while handling escape sequences. IMO, while the general structure is similar, they're distinct enough to not warrant a shared base.

Reply:
Alternatively, you can pass in a `base` of type `2 | 8 | 10 | 16` like in `scanBinaryOrOctalDigits` and special case hex characters for base 16?

Reply:
Worth keeping focus on performance when parsing numbers, even at a cost of flexibility. Too hot a code path.